src/turbomind/models/llama/SequenceManager.cc

```diff
@@ -68,7 +70,10 @@ void SequenceManager::Erase(std::map<uint64_t, Sequence>::iterator& it)
     else {
         UpdateAndSetUnlock(seq);
     }
-    freed_.insert(freed_.end(), seq.blocks.begin(), seq.blocks.end());
+    // if prefix cache enabled, blocks will be shared by sequences, cannot be freed immediately
+    if (!block_trie_->enabled()) {
+        freed_.insert(freed_.end(), seq.blocks.begin(), seq.blocks.end());
```

Comment:
> This still need to be handled, otherwise much more blocks than needed will be allocated when on demand allocation mode is used.

What's the on demand allocation mode? Now the allocation seems to only happen in `Materialize`.

Reply:
But we cannot move these blocks which were just cached to `free_` here. It's better to use allocated free blocks and then evict the cached blocks if the gpu memory is not enough.

Reply:
> When `cache_chunk_size` is set to positive value, instead of pre-allocating all blocks, the block manager will only allocate another chunk when there are not enough free blocks. Fail to release cache blocks (cached -> free) of dead sequences will make the manager allocate more blocks than needed.

@lzhangzz The Prefix Cache has been basically modified according to your suggestions. We just need to discuss this part, when would it be convenient for you to take a look? Thanks.

Reply:
@lzhangzz When Prefix Cache is enabled, we are unable to add the sequence blocks to `freed_` here. Even though we did not explicitly perform this action, the status of `freed_` will be updated by the next `Materialize`. I think it's acceptable and just a tradeoff.